pkg/helpers/sa_syncer.go

```diff
@@ -16,108 +17,140 @@ import (
 	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
 )
 
-// EnsureSAToken get the saToken of target sa and ensure it was rendered as expected.
-// A usage example combined with RenderToKubeconfigSecert would be like the following:
-// ...
-// saName := clusterManagerName+"-registration-controller-sa"
-// secretName := saName+"-kubeconfig"
-// err := helpers.EnsureSAToken(ctx, saName, clustermanagerNamespace, amdinKubeClient,
-// 	helpers.RenderToKubeconfigSecret(secretName, clustermanagerNamespace, AdminKubeconfig, kubeClient, recorder)) // kubeClient used to create secret; normally recorder is from controller object.
-// if err != nil {
-// 	return err
-// }
-// ...
-func EnsureSAToken(ctx context.Context, saName, saNamespace string, saClient kubernetes.Interface, renderSAToken func([]byte) error) error {
-	// get the service account
-	sa, err := saClient.CoreV1().ServiceAccounts(saNamespace).Get(ctx, saName, metav1.GetOptions{})
-	if err != nil {
-		return err
-	}
-	if len(sa.Secrets) == 0 {
-		return fmt.Errorf("token secret for %s not exist yet", saName)
-	}
-
-	for _, secret := range sa.Secrets {
-		// get the token secret
-		tokenSecretName := secret.Name
+type TokenGetterFunc func() ([]byte, error)
 
-		// get the token secret
-		tokenSecret, err := saClient.CoreV1().Secrets(saNamespace).Get(ctx, tokenSecretName, metav1.GetOptions{})
+// SATokenGetter get the saToken of target sa. If there is not secrets in the sa, use the tokenrequest to get a token.
+func SATokenGetter(ctx context.Context, saName, saNamespace string, saClient kubernetes.Interface) TokenGetterFunc {
+	return func() ([]byte, error) {
+		// get the service account
+		sa, err := saClient.CoreV1().ServiceAccounts(saNamespace).Get(ctx, saName, metav1.GetOptions{})
 		if err != nil {
-			return err
+			return nil, err
 		}
 
-		if tokenSecret.Type != corev1.SecretTypeServiceAccountToken {
-			continue
+		for _, secret := range sa.Secrets {
+			// get the token secret
+			tokenSecretName := secret.Name
+
+			// get the token secret
+			tokenSecret, err := saClient.CoreV1().Secrets(saNamespace).Get(ctx, tokenSecretName, metav1.GetOptions{})
+			if err != nil {
+				return nil, err
+			}
+
+			if tokenSecret.Type != corev1.SecretTypeServiceAccountToken {
+				continue
+			}
+
+			saToken, ok := tokenSecret.Data["token"]
+			if !ok {
+				continue
+			}
+
+			return saToken, nil
 		}
 
-		saToken, ok := tokenSecret.Data["token"]
-		if !ok {
-			return fmt.Errorf("no token in data for secret %s", tokenSecretName)
+		// 8640 hour
+		expirationSec := int64(8640 * 3600)
```

Comment:
what will happen after this period? will the klusterlet refresh the token?

Reply:
I add a TODO to check the validity of the token and refresh if necessary